components/campaign/CampaignChaosBagView/index.tsx

```diff
@@ -0,0 +1,363 @@
+            systemItem: 'cancel',
+            text: t`Cancel`,
+            id: 'back',
+            color: COLORS.navButton,
+            testID: t`Cancel`,
+          } : {
+            icon: iconsMap['arrow-left'],
+            id: 'androidBack',
+            color: COLORS.navButton,
+            testID: t`Back`,
+          },
+        ],
+        rightButtons: [{
+          systemItem: 'save',
+          text: t`Edit`,
+          id: 'edit',
+          showAsAction: 'ifRoom',
+          color: COLORS.navButton,
+          testID: t`Edit`,
+        }],
+      },
+    };
+  }
+
+  _navEventListener?: EventSubscription;
+
+  constructor(props: Props) {
+    super(props);
+
+    this._navEventListener = Navigation.events().bindComponent(this);
+  }
+
+  componentWillUnmount() {
+    this._navEventListener && this._navEventListener.remove();
+  }
+
+  navigationButtonPressed({ buttonId }: { buttonId: string }) {
+    const { componentId } = this.props;
+    if (buttonId === 'back' || buttonId === 'androidBack') {
+      Navigation.pop(componentId);
+    } else if (buttonId === 'edit') {
+      this._showChaosBagDialog();
+    }
+  }
+
+  _handleClearTokensPressed = () => {
+    this.clearTokens();
+  };
+
+  _handleDrawTokenPressed = () => {
+    const { chaosBagResults } = this.props;
+    if (chaosBagResults.drawnTokens.length >= 1) {
+      this.clearTokens();
+    } else {
+      this.drawToken();
+    }
+  };
+
+  _handleAddAndDrawAgainPressed = () => {
+    this.drawToken();
+  };
+
+  _handleSealTokenPressed = () => {
+  };
+
+  _showChaosBagDialog = () => {
+    const {
+      componentId,
+      chaosBag,
+      updateChaosBag,
+    } = this.props;
+
+
+    Navigation.push<EditChaosBagProps>(componentId, {
+      component: {
+        name: 'Dialog.EditChaosBag',
+        passProps: {
+          chaosBag,
+          updateChaosBag: updateChaosBag,
+          trackDeltas: true,
+        },
+        options: {
+          topBar: {
+            title: {
+              text: t`Chaos Bag`,
+            },
+            backButton: {
+              title: t`Cancel`,
+            },
+          },
+        },
+      },
+    });
+  };
+
+  getWeighedChaosBag = (list: any[], weight: any[]) => {
+    const weighedList = [];
+
+    for (let i = 0; i < weight.length; i++) {
+      const multiples = weight[i];
+
+      for (let j = 0; j < multiples; j++) {
+        weighedList.push(list[i]);
+      }
+    }
+
+    return weighedList;
+  }
+
+  getRandomChaosToken = (chaosBag: ChaosBag) => {
+    const list = keys(chaosBag);
+    const weight = values(chaosBag);
+    const weighedList = this.getWeighedChaosBag(list, weight);
+
+    return shuffle(weighedList)[0];
+  };
+
+  drawToken = () => {
+    const { campaignId, chaosBag, chaosBagResults } = this.props;
+
+    const newIconKey = this.getRandomChaosToken(chaosBag);
+    const drawnTokens = chaosBagResults.drawnTokens;
+
+    if (newIconKey) {
+      drawnTokens.push(newIconKey);
```

Comment:
So this is the problem here.  Because the drawnTokens is being directly mutated (instead of making a copy of it), the prop isn't invalidating (since its the same object as before).  You always want to make a copy of a prop to modify, never edit them directly, or it can break the way that Redux tracks invalidation (basically if its the same 'shallow' object, it doesn't look any deeper).
Using a spread operator or a slice() before doing the push will fix this.
`const drawnTokens = [...chaosBagResults.drawnTokens];`